src/Data/Bin/Properties.agda

```diff
@@ -1,143 +1,913 @@
+  where
+  m = toℕ x;  n = toℕ y;  1+n = 1+ n
+
+toℕ-homo-+ (suc2* x) (suc2* y) = begin
+  toℕ ((suc2* x) + (suc2* y))       ≡⟨⟩
+  toℕ (suc (suc2* (x + y)))         ≡⟨ toℕ∘suc (suc2* (x + y)) ⟩
+  1+ (toℕ (suc2* (x + y)))          ≡⟨⟩
+  1+ (1+ (2 *ₙ (toℕ (x + y))))      ≡⟨ cong (1+_ ∘ 1+_ ∘ (2 *ₙ_))
+                                                       (toℕ-homo-+ x y) ⟩
+  1+ (1+ (2 *ₙ (m +ₙ n)))           ≡⟨ cong (1+_ ∘ 1+_) (ℕp.*-distribˡ-+ 2 m n)
+                                     ⟩
+  1+ (1+ (2 *ₙ m +ₙ 2 *ₙ n))        ≡⟨ cong 1+_
+                                            (sym (ℕp.+-assoc 1 (2 *ₙ m) _)) ⟩
+  1+ ((1+ (2 *ₙ m) +ₙ (2 *ₙ n)))    ≡⟨ (Of+ℕ-semigroup.x∙yz≈y∙xz
+                                                         1 (1+ (2 *ₙ m)) _) ⟩
+  (1+ (2 *ₙ m)) +ₙ (1+ (2 *ₙ n))    ≡⟨⟩
+  toℕ (suc2* x) +ₙ toℕ (suc2* y)    ∎
+  where
+  m = toℕ x;  n = toℕ y
+
+
+ext-fromℕ :  (n : ℕ) → ∃ (\x → toℕ x ≡ n)    -- Mind:  it costs O(n)
+ext-fromℕ 0       =  (0# , refl)
+ext-fromℕ (1+ n)  with ext-fromℕ n
+...
+    | (x , toℕ-x≡n) = (suc x , toℕ-suc-x≡1+n)
+  where
+  toℕ-suc-x≡1+n = begin
+    toℕ (suc x)   ≡⟨ toℕ∘suc x ⟩
+    1+ (toℕ x)    ≡⟨ cong 1+_ toℕ-x≡n ⟩
+    1+ n          ∎
+
+fromℕ : ℕ → Bin
+fromℕ = proj₁ ∘ ext-fromℕ
+
+fromℕ-homo-+ :  (m n : ℕ) → fromℕ (m +ₙ n) ≡ fromℕ m + fromℕ n
+fromℕ-homo-+ 0      _ = refl
+fromℕ-homo-+ (1+ m) n = begin
+  fromℕ ((1+ m) +ₙ n)         ≡⟨⟩
+  suc (fromℕ (m +ₙ n))        ≡⟨ cong suc (fromℕ-homo-+ m n) ⟩
+  suc (a + b)                 ≡⟨ sym (suc-+ a b) ⟩
+  (suc a) + b                 ≡⟨⟩
+  (fromℕ (1+ m)) + (fromℕ n)  ∎
+  where
+  a = fromℕ m;  b = fromℕ n
+
+toℕ∘fromℕ :  toℕ ∘ fromℕ ≗ id
+toℕ∘fromℕ 0      = refl
+toℕ∘fromℕ (1+ n) = begin
+  toℕ (fromℕ (1+ n))   ≡⟨⟩
+  toℕ (suc (fromℕ n))  ≡⟨ toℕ∘suc (fromℕ n) ⟩
+  1+ (toℕ (fromℕ n))   ≡⟨ cong 1+_ (toℕ∘fromℕ n) ⟩
+  1+ n                 ∎
+
+toℕ-injective :  ∀ {x y} → toℕ x ≡ toℕ y → x ≡ y
+toℕ-injective {0#}     {0#}     _               =  refl
+toℕ-injective {2suc x} {2suc y} 2[1+xN]≡2[1+yN] =  cong 2suc x≡y
+  where
+  xN        = toℕ x
+  yN        = toℕ y
+  1+xN≡1+yN = ℕp.*-cancelˡ-≡ {1+ xN} {1+ yN} 1 2[1+xN]≡2[1+yN]
+  xN≡yN     = cong predₙ 1+xN≡1+yN
+  x≡y       = toℕ-injective xN≡yN
+
+toℕ-injective {2suc x} {suc2* y} 2[1+xN]≡1+2yN =
+                               contradiction 2[1+xN]≡1+2yN (even≢odd 1+xN yN)
+  where
+  1+xN = 1+ (toℕ x);  yN = toℕ y
+
+toℕ-injective {suc2* x} {2suc y} 1+2xN≡2[1+yN] =
+                        contradiction (sym 1+2xN≡2[1+yN]) (even≢odd 1+yN xN)
+  where
+  xN = toℕ x;  1+yN = 1+ (toℕ y)
+
+toℕ-injective {suc2* x} {suc2* y} 1+2xN≡1+2yN =  cong suc2* x≡y
+  where
+  xN = toℕ x;  yN = toℕ y;  2xN≡2yN = cong predₙ 1+2xN≡1+2yN
+
+  xN≡yN = ℕp.*-cancelˡ-≡ {xN} {yN} 1 2xN≡2yN;   x≡y = toℕ-injective xN≡yN
+
+
+toℕ-surjective :  (y : ℕ) → ∃ (\x → toℕ x ≡ y)
+toℕ-surjective 0       =  (0# , refl)
+toℕ-surjective (1+ n)  with toℕ-surjective n
+...
+    | (a , toℕ-a≡n) = (suc a , toℕ-suc-a≡1+n)
```

Comment:
Please put the rest of the case on the same line as the `...`. The equational reasoning also doesn't need to go in a `where`.

Reply:
Done.
And the eq reasoning is replaced with
```
  toℕ-suc-a≡1+n =  trans (toℕ∘suc a) (cong 1+_ toℕ-a≡n)
```